File: 03_Longest_subtring/solution.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_set<char> charSet;
        int left = 0;
        int maxLength = 0;

        for (int right = 0; right < s.length(); right++) {
            // if duplicates (the character alr exists), shrink the window from the left until the duplicate is removed
            while (charSet.find(s[right]) != charSet.end()) {
                charSet.erase(s[left]);
                // increase left or shrink the window until no more duplicates
                left++;
            }

            // if no duplicates, add the character to the set
            charSet.insert(s[right]);

            // right-left+1 counts the total elements currently in the set
            maxLength = max(maxLength, right-left+1);
        }

        return maxLength;
    }
};

File: 03_Longest_subtring/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_set = set() # faster than using list
        left = 0
        max_length = 0

        # right will automatically increments from 0 to the end of the string
        for right in range(len(s)):
            while s[right] in char_set:
                char_set.remove(s[left])
                left += 1
            
            char_set.add(s[right])

            max_length = max(max_length, right - left + 1)
        
        return max_length

        